### PR TITLE
Add VSCode command to view the hir of a function body

### PR DESCRIPTION
**Pull request by theotherphil**
Will fix https://github.com/rust-analyzer/rust-analyzer/issues/7061. Very rough initial version just to work out where I needed to wire everything up.

@matklad would you be happy merging a hir visualiser of some kind? If so, do you have any thoughts on what you'd like it show, and how?

I've spent very little time on this thus far, so I'm fine with throwing away the contents of this PR, but I want to avoid taking the time to make this more polished/interactive/useful only to discover that no-one else has any interest in this functionality.

image

### CODE REVIEW

**Comment by matklad** [comment]
>would you be happy merging a hir visualiser of some kind? If so, do you have any thoughts on what you'd like it show, and how?

yup, as long as the functionality is isolated! Don't have thoughts about specific format

File: crates/ide/src/view_hir.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    let function: Function = sema.to_def(&function)?;
+    let body = db.body(function.id.into());
+
+    let mut result = String::new();
+    writeln!(&mut result, "== Body expressions ==").ok()?;
+
+    for (id, expr) in body.exprs.iter() {
+        writeln!(&mut result, "{:?}: {:?}", id, expr).ok()?;
+    }
```
Comment (matklad) [inline crates/ide/src/view_hir.rs:36]:
```suggestion
    let res = function.debug_hir(db);
```

**Comment by theotherphil** [comment]
This is ready for review.

The output is pretty basic (it's unchanged since the original screenshot, except that it now also shows the function name), but is sufficient for my original use case.

File: crates/hir/src/code_model.rs [hunk 2]
```diff
@@ -797,6 +797,19 @@ impl Function {
     pub fn has_body(self, db: &dyn HirDatabase) -> bool {
         db.function_data(self.id).has_body
     }
+
+    /// A textual representation of the HIR of this function for debugging purposes.
+    pub fn debug_hir(self, db: &dyn HirDatabase) -> String {
+        let body = db.body(self.id.into());
+
+        let mut result = String::new();
+        writeln!(&mut result, "HIR expressions in the body of `{}`:", self.name(db)).unwrap();
```
Comment (matklad) [inline crates/hir/src/code_model.rs:806]:
```suggestion
        format_to!(result, "HIR expressions in the body of `{}`:", self.name(db));
```

**Review by matklad — COMMENTED** [review]
r=me with format_to fix

bors d+

**Pull request merged**